src/Featurizers/Structs.h

```diff
@@ -95,6 +95,65 @@ class SingleValueSparseVectorEncoding {
     bool operator!=(SingleValueSparseVectorEncoding const &other) const;
 };
 
+
+/////////////////////////////////////////////////////////////////////////
+
+///  \class        CSREncoding
+///  \brief         Struct that contains information that can be used to
+///                 create a sparse matrix without creating the matrix
+///                 itself. We are using Compressed Sparse Row form
+///                 in this case.
+///
+///                 CSREncoding contains three vectors to express a
+///                 sparse matrix:
+///                 V contains all non-zero element values,
+///                 ColIndex contains column indices of non-zero values,
+///                 RowIndex has one element per row in the matrix and encodes
+///                 the index in V where the given row starts
+///
+
+template <typename T>
+class CSREncoding {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using value_type                        = T;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Data
+    // |
+    // ----------------------------------------------------------------------
+
+    std::vector<value_type>          Values;
+    std::vector<std::uint64_t>       ColIndex;
+    std::vector<std::uint64_t>       RowIndex;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    CSREncoding();
+    CSREncoding(std::vector<value_type> values, std::vector<std::uint64_t> col, std::vector<std::uint64_t> row);
+
+    bool operator==(CSREncoding const &other) const;
+    bool operator!=(CSREncoding const &other) const;
+};
+
+/////////////////////////////////////////////////////////////////////////
+///  \enum          NormType
+///  \brief         Training state associated with an `Estimator`.
+///
+enum class NormType : unsigned char {

```

Comment:
>NormType [](start = 11, length = 8)

This doesn't belong in this file - move this definition closer to where it is actually used.